storage/bootstrap.go

```diff
@@ -33,37 +36,111 @@ const (
 	bootstrapped
 )
 
-func analyze(bs bootstrapState, entity string) (bool, error) {
-	switch bs {
-	case bootstrapped:
-		return false, nil
-	case bootstrapping:
-		return false, fmt.Errorf("%s is being bootstrapped", entity)
-	default:
-		return true, nil
+var (
+	// errDatabaseIsBootstrapping raised when trying to bootstrap a database that's being bootstrapped.
+	errDatabaseIsBootstrapping = errors.New("database is bootstrapping")
+
+	// errDatabaseNotBootstrapped raised when trying to query a database that's not yet bootstrapped.
+	errDatabaseNotBootstrapped = errors.New("database is not yet bootstrapped")
+
+	// errShardIsBootstrapping raised when trying to bootstrap a shard that's being bootstrapped.
+	errShardIsBootstrapping = errors.New("shard is bootstrapping")
+
+	// errShardNotBootstrapped raised when trying to flush data to disk for a shard that's not yet bootstrapped.
+	errShardNotBootstrapped = errors.New("shard is not yet bootstrapped")
+
+	// errSeriesIsBootstrapping raised when trying to bootstrap a series that's being bootstrapped.
+	errSeriesIsBootstrapping = errors.New("series is bootstrapping")
+
+	// errSeriesNotBootstrapped raised when trying to flush data to disk for a series that's not yet bootstrapped.
+	errSeriesNotBootstrapped = errors.New("series is not yet bootstrapped")
+)
+
+// bootstrapManager manages the bootstrap process.
+type bootstrapManager interface {
+	// IsBootstrapped returns whether the database is already bootstrapped.
+	IsBootstrapped() bool
+
+	// Bootstrap performs bootstrapping for all shards owned by db. It returns an error
+	// if the server is currently being bootstrapped, and nil otherwise.
+	Bootstrap() error
+}
+
+type bsManager struct {
+	sync.RWMutex
+
+	database *db                  // storage database
```

Comment:
Can we avoid relying on the concrete `db` struct and rely on the interface itself?  This way we can test the component in complete isolation and always the dependencies, which is the pattern I've been trying to keep amongst all the code so far.